### ISSUE TITLE
Cannot infer type of lambda regression with mypy 1.9

### ISSUE THREAD

**Issue by rbuffat** [python/mypy#17018]
**Bug Report**

The lambda in the following code snippet reports `error: Cannot infer type of lambda` starting from mypy 1.9.0 (It did not report an error with mypy 1.8.0).

https://github.com/python/mypy/issues/15459 seems related, but it looks as  mypy 1.8.0 also reported an error there.

**To Reproduce**

```python
import asyncio
import signal


async def shutdown(
    signal: signal.Signals,
) -> None:
    print(signal.name)


async def main() -> None:
    loop = asyncio.get_event_loop()
    signals = (signal.SIGHUP, signal.SIGTERM, signal.SIGINT)
    for s in signals:
        loop.add_signal_handler(
            s,
            lambda s=s: asyncio.create_task(  #  <-- error: Cannot infer type of lambda  [misc]
                shutdown(
                    signal=s,
                ),
                name="shutdown",
            ),
        )
```

[mypy playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=5d5f254041ca43e9dc9c21a86394559c)

**Expected Behavior**

No error shown regarding the lambda.

**Actual Behavior**

mypy reports Cannot infer type of lambda  [misc]

**Your Environment**

- Mypy version used: 1.9.0
- Mypy command-line flags: --strict
- Mypy configuration options from `mypy.ini` (and other config files): -
- Python version used: 3.11.8

**Comment by AlexWaygood** [comment]
Here's a mimimal repro of the underlying issue that doesn't involve typeshed's stubs for asyncio:

```py
from typing import TypeVarTuple, Callable

x = 42
Ts = TypeVarTuple("Ts")

def foo(cb: Callable[[*Ts], object], *args: *Ts) -> None: ...
foo(lambda x=x: None)  # error: Cannot infer type of lambda  [misc]
```

https://mypy-play.net/?mypy=latest&python=3.12&gist=ce433643f1bff8c03f6f8149f7829793

(The immediate cause of the new false positive was https://github.com/python/typeshed/commit/a08d2d5d6693d9a53f063b6a495a23f1bff7d010, but the underlying issue is mypy struggling to solve a TypeVarTuple when you're passing in a `lambda`.)
